teps/0117-concurrency-controls.md

```diff
@@ -0,0 +1,99 @@
+---
+status: proposed
+title: Concurrency Controls
+creation-date: '2022-08-19'
+last-updated: '2022-08-19'
+authors:
+- '@vdemeester'
+- '@williamlfish'
+- '@lbernick'
+---
+
+# TEP-0117: Concurrency Controls
+
+<!-- toc -->
+- [Summary](#summary)
+- [Motivation](#motivation)
+  - [Use Cases](#use-cases)
+  - [Requirements](#requirements)
+  - [Future Work](#future-work)
+- [Proposal](#proposal)
+- [References](#references)
+<!-- /toc -->
+
+## Summary
+
+Allow users to configure desired behavior for managing concurrent PipelineRuns.
+
+## Motivation
+
+Allow users to avoid wasting resources on redundant work, to prevent concurrent runs of
+non-idempotent operations, and to have more control over concurrent load on a cluster.
+
+### Use Cases
+
+Avoiding redundant work
+- A developer pushes a pull request and quickly notices and fixes a typo. They would like to have the first CI run automatically canceled and replaced by the second.
+
+Controlling non-idempotent operations
+- An organization uses a Pipeline for deployment, and wants to make sure that only the most recent changes are applied to their cluster.
+If a new deployment PipelineRun starts while a previous one is still running, they would like the previous one to be canceled.
+  - The organization might want only one deployment PipelineRun per cluster, per namespace, per environment (prod/staging),
+  per repo, or per user ([example](https://github.com/tektoncd/pipeline/issues/2828#issuecomment-646150534)).
+- An integration test communicates with a stateful external service (like a database), and a developer wants to ensure
+that integration testing TaskRuns within their CI PipelineRun don’t run concurrently with other integration testing TaskRuns
+(based on [this comment](https://github.com/tektoncd/pipeline/issues/2828#issuecomment-647747330)).
+- An organization would like to ensure Terraform operations run serially (based on [this comment](https://github.com/tektoncd/experimental/issues/699#issuecomment-951606279))
+  - This user would like to be able to cancel queued runs that are pending, rather than cancelling running runs. 
+
+Controlling load on a cluster
+- An organization has multiple teams working on a mobile application with a limited number of test devices.
+They want to limit the number of concurrent CI runs per team, to prevent one team from using all the available devices
+and crowding out CI runs from other teams.
+- A Pipeline performs multiple parallelizable things with different concurrency caps, as described in [this comment](https://github.com/tektoncd/pipeline/issues/2591#issuecomment-626778025).
+- Allow users to cap the number of matrixed TaskRuns (alpha) that can run at a given time.
```

Comment:
@jerop I've included the matrix concurrency use case here as one we want to try to solve for, but I think for an initial version of this proposal it'll be easier to focus just on concurrency of PipelineRuns and descope TaskRuns for future work.

Reply:
@lbernick note that matrix is used for taskruns and runs for now but it'll be available for pipelineruns as well once pipeline in pipelines is implemented -- so it's still relevant if the focus is pipelineruns